remotes/docker/registry.go

```diff
@@ -202,11 +204,29 @@ func MatchAllHosts(string) (bool, error) {
 // MatchLocalhost is a host match function which returns true for
 // localhost.
 func MatchLocalhost(host string) (bool, error) {
-	for _, s := range []string{"localhost", "127.0.0.1", "[::1]"} {
-		if len(host) >= len(s) && host[0:len(s)] == s && (len(host) == len(s) || host[len(s)] == ':') {
-			return true, nil
-		}
+	switch {
+	// host name of the form localhost with an
+	// optional port number
+	case strings.HasPrefix(host, "localhost"):
+		return len(host) == len("localhost") ||
+			host[len("localhost")] == ':', nil
+
+	// IPs of the form [::1] with an optional
+	// port number
+	case strings.HasPrefix(host, "[::1]"):
+		return len(host) == len("[::1]") ||
+			host[len("[::1]")] == ':', nil
+
+	// IPs in the range 127.0.0.0 to 127.255.255.255
+	// with an optional port number
+	case strings.HasPrefix(host, "127"):
```

Comment:
Please use https://golang.org/src/net/ipsock.go?s=4766:4832#L146

It should be also noted that 0x7F000001 is a valid loopback IP.

Reply:
Thank you, I will look into `SplitHostPort` and make the necessary changes!

Reply:
> It should be also noted that "0x7F000001" is a valid loopback IP.

Hey, I was thinking about how to go about doing this, the `net` package provides utility to go from an IP to its hex form, but not the reverse way. Here's what I'm thinking:
- For hex IP cases, use `LookupIP` which returns a slice of `netIP` consisting of IPv4 and IPv6 addresses of the host. 
    - Before using `LookupIP` on it, a `SplitHostPort` might have to be done. And from the result of this, use `IsLoopback()`

Does this sound okay?